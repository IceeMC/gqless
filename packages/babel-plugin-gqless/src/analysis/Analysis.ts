import { types as t, NodePath } from '@babel/core'
import { Cache, FileAnalysis } from '.'

export class Analysis {
  public file!: FileAnalysis
  public cache!: Cache

  constructor(public parent?: Analysis) {
    if (parent) {
      this.file = parent.file
      this.cache = parent.cache
    }
  }
}
