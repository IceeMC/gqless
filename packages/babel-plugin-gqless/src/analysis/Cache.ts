import { FileAnalysis } from '.'

export class Cache {
  public files = new Map<string, FileAnalysis>()

  constructor(private api: typeof import('@babel/core')) {}

  public getPath(path: string) {
    if (this.files.has(path)) {
      return this.files.get(path)!
    }

    const analysis = new FileAnalysis(this, path)
    this.files.set(path, analysis)
    return analysis
  }

  public transform(path: string) {
    return this.api.transformFileSync(path, {
      ast: true,
    })!.ast!
  }
}
