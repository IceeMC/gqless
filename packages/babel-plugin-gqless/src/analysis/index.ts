export * from './Cache'
export * from './Analysis'
export * from './PropAnalysis'
export * from './FieldAnalysis'
export * from './ParamAnalysis'
export * from './FileAnalysis'
export * from './FunctionAnalysis'
