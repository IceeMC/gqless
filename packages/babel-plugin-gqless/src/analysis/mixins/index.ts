export * from './Fields'
export * from './Referable'
