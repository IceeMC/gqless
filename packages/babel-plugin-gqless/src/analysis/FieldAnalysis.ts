import { types as t, NodePath } from '@babel/core'
import { Analysis } from './Analysis'
import { Fields } from './mixins'
import { Mix } from 'mix-classes'

export class FieldAnalysis extends Mix(Analysis, Fields) {
  constructor(
    parent: Analysis & Fields,
    public name: string | 0,
    public variables?: object
  ) {
    super([parent])
  }
}
