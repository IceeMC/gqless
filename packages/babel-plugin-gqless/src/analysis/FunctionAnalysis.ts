import { types as t, NodePath } from '@babel/core'
import { Analysis } from './Analysis'
import { ParamAnalysis } from './ParamAnalysis'
import { Referable } from './mixins'
import { Mix, Generic as g } from 'mix-classes'

export interface FunctionAnalysis extends Referable<t.Function> {}

export class FunctionAnalysis extends Mix(Analysis, g(Referable)) {
  public params = new Map<number, ParamAnalysis>()

  constructor(parent: Analysis, path: NodePath<t.Function>) {
    super([parent], [path])
  }

  public getParam(
    path: NodePath<
      t.Identifier | t.Pattern | t.RestElement | t.TSParameterProperty
    >
  ) {
    for (const [i, paramAnalysis] of this.params) {
      if (i === path.key) return paramAnalysis
    }

    const paramAnalysis = new ParamAnalysis(this, path)
    this.params.set(path.key as number, paramAnalysis)
    return paramAnalysis
  }

  public scan(...args: NodePath<t.Expression>[]) {
    const params = this.path.get('params')

    args.forEach((arg, i) => {
      const param = params[i]
      if (!param) return

      this.getParam(param).scan()
    })

    if (this.parent !== this.parent!.file) {
      console.log('add analysis to prev analysis')
    }
  }
}
