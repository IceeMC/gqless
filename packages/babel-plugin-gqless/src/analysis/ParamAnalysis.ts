import { types as t, NodePath } from '@babel/core'
import { FunctionAnalysis } from './FunctionAnalysis'
import { Fields, Referable } from './mixins'
import { Analysis } from './Analysis'
import { Mix, Generic as g } from 'mix-classes'
import { objectPropValue, evalProperty } from '../utils'

type Param = t.Identifier | t.Pattern | t.RestElement | t.TSParameterProperty

export interface ParamAnalysis extends Referable<Param> {}

export class ParamAnalysis extends Mix(Analysis, g(Referable), Fields) {
  constructor(parent: FunctionAnalysis, path: NodePath<Param>) {
    super([parent], [path])
  }

  public scan() {
    // function ({ })
    if (this.path.isObjectPattern()) {
      for (const prop of this.path.get('properties')) {
        // function ({ arg })
        if (prop.isObjectProperty()) {
          const binding = this.path.scope.getBinding(objectPropValue(prop))!

          const fieldName = evalProperty(prop)
          if (fieldName === undefined) return

          const fieldAnalysis = this.getField(fieldName)
          for (const refPath of binding.referencePaths) {
            fieldAnalysis.scanField(refPath)
          }
        }

        // function ({ ...rest })
        if (prop.isRestElement()) {
          const binding = this.path.scope.getBinding(
            (prop.node.argument as t.Identifier).name
          )!

          for (const refPath of binding.referencePaths) {
            this.scanField(refPath)
          }
        }
      }
    }

    // function (arg)
    if (this.path.isIdentifier()) {
      const binding = this.path.scope.getBinding(this.path.node.name)!

      for (const refPath of binding.referencePaths) {
        this.scanField(refPath)
      }
    }
  }
}
