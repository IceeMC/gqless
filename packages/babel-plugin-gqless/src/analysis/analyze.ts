import { dirname } from 'path'
import { types as t, NodePath } from '@babel/core'
import { invariant } from '@gqless/utils'
import { findModule, evaluate, evalAsString, evalProperty } from '../utils'
import { Analysis } from './Analysis'
import { FunctionAnalysis } from './FunctionAnalysis'
import { PropAnalysis } from './PropAnalysis'
import { ParamAnalysis } from './ParamAnalysis'

const shouldAnalyze = (
  filter: NodePath<t.Node>,
  test: NodePath<t.Node>
): boolean => {
  if (filter.isObjectExpression()) {
    const propToTest = evalAsString(test)
    if (!propToTest) return true

    for (const prop of filter.get('properties')) {
      if (prop.isObjectProperty()) {
        if (evalAsString(prop) === propToTest) return true
      }
    }

    return false
  }

  // if (
  //   filter.isNullLiteral() ||
  //   (filter.isIdentifier() && filter.node.name === 'undefined')
  // ) {
  //   return false
  // }

  return true
}

const analyzeImport = (
  { file, cache }: Analysis,
  path: NodePath<
    t.ImportDefaultSpecifier | t.ImportSpecifier | t.ImportNamespaceSpecifier
  >,
  args: NodePath<t.Expression>[]
) => {
  const getExportName = (): string | void => {
    if (path.isImportSpecifier()) {
      return path.node.imported.name
    }
    if (path.isImportDefaultSpecifier()) {
      return 'default'
    }
    if (path.isImportNamespaceSpecifier()) {
      const [preloadPath] = args
      invariant(preloadPath.isMemberExpression())
      invariant(t.isIdentifier(preloadPath.node.property))
      return preloadPath.node.property.name
    }
  }
  const exportName = getExportName()
  if (!exportName) return

  const modulePath = findModule(
    dirname(file.path),
    (<t.ImportDeclaration>path.parent).source.value
  )

  const moduleAnalysis = cache.getPath(modulePath)
  moduleAnalysis.getExport(exportName)
}

const analyzeVariable = (
  analysis: Analysis,
  path: NodePath<t.VariableDeclarator>,
  args: NodePath<t.Expression>[]
): Analysis | void => {
  const init = path.get('init')

  if (init.isFunctionExpression() || init.isArrowFunctionExpression()) {
    return analyzeFunction(analysis, init, args)
  }
}

const analyzeMemberExpression = (
  analysis: PropAnalysis | ParamAnalysis,
  path: NodePath<t.MemberExpression>,
  arg: NodePath<t.Expression>
) => {
  const property = evalAsString(path, evalProperty)
  if (property === undefined) return

  let variables: any

  if (path.parentPath.isCallExpression()) {
    const [varsPath] = path.parentPath.get('arguments')
    variables = evaluate(varsPath)
  }

  const propAnalysis = analysis.getProperty(property, variables)
  analyzeProp(propAnalysis, path, arg)
}

const analyzeParam = (
  analysis: PropAnalysis | ParamAnalysis,
  path: NodePath<
    t.Identifier | t.Pattern | t.RestElement | t.TSParameterProperty
  >,
  arg: NodePath<t.Expression>
) => {
  invariant(path.isIdentifier())
  const binding = path.scope.getBinding(path.node.name)!

  for (const refPath of binding.referencePaths) {
    if (refPath.parentPath.isMemberExpression()) {
      analyzeMemberExpression(analysis, refPath.parentPath, arg)
    }
  }
}

const analyzeProp = (
  analysis: PropAnalysis | ParamAnalysis,
  path: NodePath<t.MemberExpression>,
  arg: NodePath<t.Expression>
) => {
  if (path.parentPath.isMemberExpression()) {
    return analyzeMemberExpression(analysis, path.parentPath, arg)
  }
}

const analyzeFunction = (
  analysis: Analysis,
  path: NodePath<
    t.FunctionDeclaration | t.FunctionExpression | t.ArrowFunctionExpression
  >,
  args: NodePath<t.Expression>[]
) => {
  const funcAnalysis = analysis.getFunction(path)
  const params = path.get('params')

  args.slice(1).forEach((arg, i) => {
    const param = params[i]
    if (!param || !shouldAnalyze(arg, param)) return

    analyzeParam(funcAnalysis.getParam(param), param, arg)
  })

  const funcBody = path.get('body')

  if (funcBody.isBlockStatement()) {
    const blockBody = funcBody.get('body')

    console.log({ analysis, blockBody, args, params })
  }
  return funcAnalysis
}

export const analyze = (
  analysis: Analysis,
  path: NodePath<t.Node>,
  /**
   * Arguments specified to preload call
   * [funcToPreload, ...argsForFunc]
   */
  args: NodePath<t.Expression>[] = []
) => {
  if (path.parentPath.isImportDeclaration()) {
    return analyzeImport(analysis, path as any, args)
  }

  if (path.isVariableDeclarator()) {
    return analyzeVariable(analysis, path, args)
  }

  if (path.isFunctionDeclaration()) {
    return analyzeFunction(analysis, path, args)
  }
}
