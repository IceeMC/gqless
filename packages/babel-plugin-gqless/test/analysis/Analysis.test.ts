import { fileAnalysis, scan } from '../setup'
import { FunctionAnalysis, FileAnalysis } from '../../src/analysis'

describe('resolves', () => {
  describe('exports >', () => {
    test('export default', () => {
      const analysis = fileAnalysis({
        a: `
          export default () => {}
        `,
      })

      expect(analysis.getExport('default')).toBeInstanceOf(FunctionAnalysis)
    })

    test('export {x}', () => {
      const analysis = fileAnalysis({
        a: `
          const x = () => {}
          export { x }
        `,
      })

      expect(analysis.getExport('x')).toBeInstanceOf(FunctionAnalysis)
    })

    test('export const x', () => {
      const analysis = fileAnalysis({
        a: `
          export const a = () => {}
        `,
      })

      expect(analysis.getExport('a')).toBeInstanceOf(FunctionAnalysis)
    })
  })

  describe('imports >', () => {
    it('import {x}', () => {
      const analysis = fileAnalysis({
        a: `
          import { f as b } from 'b'
          export const f = b
        `,
        b: `
          export const f = () => {}
        `,
      })

      expect(analysis.getExport('f')).toBeInstanceOf(FunctionAnalysis)
    })

    it('import x', () => {
      const analysis = fileAnalysis({
        a: `
          import f from 'b'
          export const b = f
        `,
        b: `
          export default () => {}
        `,
      })

      expect(analysis.getExport('b')).toBeInstanceOf(FunctionAnalysis)
    })

    it('import * as x', () => {
      const analysis = fileAnalysis({
        a: `
          import * as b from 'b'
          export const f = b.f
        `,
        b: `
          export const f = () => {}
        `,
      })

      expect(analysis.getExport('f')).toBeInstanceOf(FunctionAnalysis)
    })
  })

  describe('re-exports >', () => {
    it('export {x} from', () => {
      const analysis = fileAnalysis({
        a: `
          export { f as x } from 'b'
        `,
        b: `
          export const f = () => {}
        `,
      })

      expect(analysis.getExport('x')).toBeInstanceOf(FunctionAnalysis)
    })

    it('export * from', () => {
      const analysis = fileAnalysis({
        a: `
          export * from 'b'
        `,
        b: `
          export const x = () => {}
        `,
      })

      expect(analysis.getExport('x')).toBeInstanceOf(FunctionAnalysis)
    })

    it('export x from', () => {
      const analysis = fileAnalysis({
        a: `
          export x from 'b'
        `,
        b: `
          export default () => {}
        `,
      })

      expect(analysis.getExport('x')).toBeInstanceOf(FunctionAnalysis)
    })

    it('export * as x from', () => {
      const analysis = fileAnalysis({
        a: `
          export * as x from 'b'
        `,
        b: `
          export const f = () => {}
        `,
      })

      expect(analysis.getExport('x')).toBeInstanceOf(FileAnalysis)
    })
  })
})

describe('evaluates values', () => {
  it('with recursive de-structuring', () => {
    const analysis = fileAnalysis({
      a: `
        const { b } = { b: { a: () => {} } }

        export const { a } = b
      `,
    })

    expect(analysis.getExport('a')).toBeInstanceOf(FunctionAnalysis)
  })

  it('with object de-structuring', () => {
    const analysis = fileAnalysis({
      a: `
        const n = 'n'
        const fn = () => {}

        export const { ['f'+n]: a } = { fn }
      `,
    })

    expect(analysis.getExport('a')).toBeInstanceOf(FunctionAnalysis)
  })

  it('with array de-structuring', () => {
    const analysis = fileAnalysis({
      a: `
        export const [, a] = [, () => {}]
      `,
    })

    expect(analysis.getExport('a')).toBeInstanceOf(FunctionAnalysis)
  })
})

describe('scans', () => {
  it('arg destructuring', () => {
    const files = {
      a: `
        export default ({ user }) => {
          user.name
        }
      `,
    }

    expect(scan(files, 'props')).toMatchSnapshot()
  })

  describe('imports >', () => {
    const scanImp = (source: string) =>
      scan(
        {
          a: source,
          b: `
              export const b = u => u.n
              export default b
            `,
        },
        'props'
      )

    it('import x', () => {
      expect(
        scanImp(`
          import x from 'b'
          export default x
        `)
      ).toMatchSnapshot()
    })

    it('import {x}', () => {
      expect(
        scanImp(`
          import { b as _ } from 'b'
          export default _
        `)
      ).toMatchSnapshot()
    })

    it('import * as x', () => {
      expect(
        scanImp(`
          import * as b from 'b'
          export default b.default
        `)
      ).toMatchSnapshot()
    })
  })

  describe('function calls', () => {
    it('with object arguments', () => {
      const files = {
        a: `
          const a = ({ u }) => u.age
          export default u => {
            const { args } = { args: { u } }
            a(args)
            u.name
          }
        `,
      }

      expect(scan(files, 'props')).toMatchInlineSnapshot(`
        FunctionAnalysis (
          0 -> {
            age {}
            name {}
          }
        )
      `)
    })

    it('with multiple arguments', () => {
      const files = {
        a: `
          const a = (a, b, c) => {
            a.a
            b.b
            c.c
          }

          export default u => {
            a(u, u, {})
            u.name
          }
        `,
      }

      expect(scan(files, 'props')).toMatchSnapshot()
    })
  })
})
