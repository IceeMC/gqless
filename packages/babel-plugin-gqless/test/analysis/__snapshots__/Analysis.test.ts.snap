// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scans arg destructuring 1`] = `
FunctionAnalysis (
  0 -> {
    user {
      name {}
    }
  }
)
`;

exports[`scans function calls with multiple arguments 1`] = `
FunctionAnalysis (
  0 -> {
    a {}
    b {}
    name {}
  }
)
`;

exports[`scans imports > import * as x 1`] = `
FunctionAnalysis (
  0 -> {
    n {}
  }
)
`;

exports[`scans imports > import {x} 1`] = `
FunctionAnalysis (
  0 -> {
    n {}
  }
)
`;

exports[`scans imports > import x 1`] = `
FunctionAnalysis (
  0 -> {
    n {}
  }
)
`;
